### PR TITLE
Allow non-default string types in all metadata calls

### PR DESCRIPTION
**Pull request by d-w-moore**
Here we patch the ModAVUMetadata interface in the Python client to allow both Unicode and bytestring fields everywhere, regardless of which string type is the norm for the running Python interpreter.

Included also:
   - a test to ensure metadata queries, also, can match bytestring and Unicode equally well
   - an extra test that creates one data object with a unicode name, but queries for it using the bytestring equivalent.

### CODE REVIEW

**Comment by trel** [comment]
this PR includes 443 NotLike... but no test?    perhaps you want that 1 commit and a test in a different PR?

rest seems good if the tests are passing.

**Comment by d-w-moore** [comment]
Will do the separate PR, I think.  NotLike is used in one of these new tests but it could use its own....

**Comment by trel** [comment]
Ah.  Okay.  Yes.  Let’s get it in first with its own test.

Then this work… which can also use the NotLike.

**Comment by d-w-moore** [comment]
Still ready for review, though.

**Comment by d-w-moore** [comment]
tests pass.

**Comment by trel** [comment]
nothing jumps out as concerning to me.   feels like it's squash time.

**Comment by d-w-moore** [comment]
Ok, squashed.

**Comment by trel** [comment]
@alanking, thoughts?

File: irods/test/meta_test.py [hunk 1]
```diff
@@ -25,6 +44,63 @@ class TestMeta(unittest.TestCase):
     attr0, value0, unit0 = 'attr0', 'value0', 'unit0'
     attr1, value1, unit1 = 'attr1', 'value1', 'unit1'
 
+    def test_stringtypes_in_general_query(self):
```
Comment (alanking) [inline irods/test/meta_test.py:47]: Issue numbers on these test names may be useful

**Review by alanking — COMMENTED** [review]
Overall seems fine. I had a couple of thoughts, but they're not really big concerns

File: irods/test/query_test.py [hunk 2]
```diff
@@ -540,6 +540,12 @@ def test_rules_query__267(self):
         if results:
             Rule(self.sess).remove_by_id( results[0][RuleExec.id] )
 
+    def test_utf8_equality_in_query(self):
+        name = u'prefix-\u1000-suffix'
+        data = self.sess.data_objects.create(u'{}/{}'.format(self.coll_path, name))
```
[the remark this answers is not in the document]
Comment (alanking) [reply irods/test/query_test.py:545]: Very good. Resolving.

File: irods/test/meta_test.py [hunk 3]
```diff
@@ -8,15 +8,34 @@
...
+def normalize_to_bytes(string, unicode_encoding='utf8'):
+    # Python2 and 3 enumerate bytestrings differently.
+    ord_ = (lambda _:_) if any(x for x in string[:1] if type(x) is int) else ord
+    if not string or ord_(max(x for x in string)) > 255:
+        return string.encode(unicode_encoding)
+    # str->bytes type conversion using element-wise copy, aka a trivial encode.
+    array = bytearray(ord_(x) for x in string)
+    return bytes(array)
+
+def equivalent_over_unicode(avu1, avu2, key = normalize_to_bytes):
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/test/meta_test.py:30]: Equal over the domain of UTF8 encodable strings is probably more exact.  Less math-y terminology might be : they map (by whatever method is appropriate, encoding or leaving as-is) to the same byte sequence.  But that's less concise ; )

[the remark this answers is not in the document]
Comment (alanking) [reply irods/test/meta_test.py:30]: Gotcha. Feel free to resolve if you're happy with the name

[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/test/meta_test.py:30]: hm ...  does `resolving_to_identical_bytestrings `seem any better?

[the remark this answers is not in the document]
Comment (trel) [reply irods/test/meta_test.py:30]: `resolves_to_identical_bytestrings` ?

[the remark this answers is not in the document]
Comment (alanking) [reply irods/test/meta_test.py:30]: No, I think what you have is fine actually. Thinking about it as a mapping function cleared it up for me.

[the remark this answers is not in the document]
Comment (alanking) [reply irods/test/meta_test.py:30]: But also, prefer `resolves_to_identical_bytestrings` to `resolving_to_identical_bytestrings`, yes.

File: irods/message/__init__.py [hunk 4]
```diff
@@ -154,6 +155,7 @@ def ET(xml_type = (), server_version = None):
     # Python 3
     UNICODE = str
 
+_METADATA_FIELD_TYPE_OPTIONS = {str,UNICODE,bytes}
```
[the remark this answers is not in the document]
Comment (d-w-moore) [reply irods/message/__init__.py:158]: I agree, will leave out the final _OPTIONS part

**Comment by alanking** [comment]
Changes look good. Please resolve open conversations and squash

File: irods/test/meta_test.py [hunk 5]
```diff
@@ -25,6 +44,63 @@ class TestMeta(unittest.TestCase):
...
+
+        # Test that application of unicode and bytestring metadata were equivalent
+        self.assertEqual(metadata[0], metadata[1])
+
+    @unittest.skipIf(PY3, 'Unicode strings are normal on Python3')
+    def test_unicode_AVUs_in_Python2__issue_442(self):
+        data_object = self.sess.data_objects.get(self.obj_path)
+        meta_set = iRODSMeta(u'\u1000', u'value', u'units')
+        meta_add = iRODSMeta(*tuple(meta_set))
+        meta_add.name += u"-add";
```
Comment (alanking) [inline irods/test/meta_test.py:85]: Semicolon

File: irods/test/meta_test.py [hunk 6]
```diff
@@ -25,6 +44,63 @@ class TestMeta(unittest.TestCase):
...
+        for index, meta in [(m.name, m) for m in (meta_add, meta_set)]:
+            fetched = data_object.metadata[index]
+            self.assertTrue(resolves_to_identical_bytestrings(fetched, meta), 'fetched unexpected meta for %r' % index)
+
+    @unittest.skipIf(PY2, 'Byte strings are normal on Python2')
+    def test_bytestring_AVUs_in_Python3__issue_442(self):
+        data_object = self.sess.data_objects.get(self.obj_path)
+        meta_set = iRODSMeta(u'\u1000'.encode('utf8'),b'value',b'units')
+        meta_add = iRODSMeta(*tuple(meta_set))
+        meta_add.name += b"-add";
```
Comment (alanking) [inline irods/test/meta_test.py:97]: Semicolon

**Review by alanking — COMMENTED** [review]
Hmm... sorry for the back and forth. Just took a look at Codacy and it seems to raise a couple of valid points. And of course we will need to get #445 in before this, as you said before.

**Comment by d-w-moore** [comment]
Yup, no biggie... will take care of those.

**Review by trel — APPROVED** [review]

**Review by alanking — APPROVED** [review]
Looks good. #445 is in now. Please squash and #.
